commit 4c79ec09e982937029d0b59be331a47d8c2449f7
Author: Grace Do <xgrace@gmail.com>
Date:   Mon Mar 29 22:19:57 2021 -0700

    Avoid setting fields added to CRDs across upgrade

    The upgrade from 9.x to 12.x added new fields to the Prometheus and Alertmanager
    V1 CRDs. This causes upgrades to fail due to validation errors.

    To avoid upgrade errors, we add a new "includeNewCRDFields" value which
    is used to determine whether these new fields should be set or not.
    It is defaulted to false to allow the initial upgrade to 12.x to succeed.
    Users should be advised to override/set this to true after upgrading
    if they wish to use the affected fields. The field should be removed from
    the chart in the next major release.

diff --git a/staging/prometheus-operator/templates/alertmanager/alertmanager.yaml b/staging/prometheus-operator/templates/alertmanager/alertmanager.yaml
index 1115e6145..7caef2c8b 100644
--- a/staging/prometheus-operator/templates/alertmanager/alertmanager.yaml
+++ b/staging/prometheus-operator/templates/alertmanager/alertmanager.yaml
@@ -44,18 +44,22 @@ spec:
   configMaps:
 {{ toYaml .Values.alertmanager.alertmanagerSpec.configMaps | indent 4 }}
 {{- end }}
+{{- if .Values.includeNewCRDFields }}
 {{- if .Values.alertmanager.alertmanagerSpec.alertmanagerConfigSelector }}
   alertmanagerConfigSelector:
 {{ toYaml .Values.alertmanager.alertmanagerSpec.alertmanagerConfigSelector | indent 4}}
 {{ else }}
   alertmanagerConfigSelector: {}
 {{- end }}
+{{- end }}
+{{- if .Values.includeNewCRDFields }}
 {{- if .Values.alertmanager.alertmanagerSpec.alertmanagerConfigNamespaceSelector }}
   alertmanagerConfigNamespaceSelector:
 {{ toYaml .Values.alertmanager.alertmanagerSpec.alertmanagerConfigNamespaceSelector | indent 4}}
 {{ else }}
   alertmanagerConfigNamespaceSelector: {}
 {{- end }}
+{{- end }}
 {{- if .Values.alertmanager.alertmanagerSpec.resources }}
   resources:
 {{ toYaml .Values.alertmanager.alertmanagerSpec.resources | indent 4 }}

diff --git a/staging/prometheus-operator/templates/prometheus/prometheus.yaml b/staging/prometheus-operator/templates/prometheus/prometheus.yaml
index cf843c274..29f4c2797 100644
--- a/staging/prometheus-operator/templates/prometheus/prometheus.yaml
+++ b/staging/prometheus-operator/templates/prometheus/prometheus.yaml
@@ -65,7 +65,9 @@ spec:
 {{- end }}
   paused: {{ .Values.prometheus.prometheusSpec.paused }}
   replicas: {{ .Values.prometheus.prometheusSpec.replicas }}
+{{- if .Values.includeNewCRDFields }}
   shards: {{ .Values.prometheus.prometheusSpec.shards }}
+{{- end }}
   logLevel:  {{ .Values.prometheus.prometheusSpec.logLevel }}
   logFormat:  {{ .Values.prometheus.prometheusSpec.logFormat }}
   listenLocal: {{ .Values.prometheus.prometheusSpec.listenLocal }}
@@ -134,6 +136,7 @@ spec:
 {{ else }}
   podMonitorNamespaceSelector: {}
 {{- end }}
+{{- if .Values.includeNewCRDFields }}
 {{- if .Values.prometheus.prometheusSpec.probeSelector }}
   probeSelector:
 {{ toYaml .Values.prometheus.prometheusSpec.probeSelector | indent 4 }}
@@ -144,12 +147,15 @@ spec:
 {{ else }}
   probeSelector: {}
 {{- end }}
+{{- end }}
+{{- if .Values.includeNewCRDFields }}
 {{- if .Values.prometheus.prometheusSpec.probeNamespaceSelector }}
   probeNamespaceSelector:
 {{ toYaml .Values.prometheus.prometheusSpec.probeNamespaceSelector | indent 4 }}
 {{ else }}
   probeNamespaceSelector: {}
 {{- end }}
+{{- end }}
 {{- if .Values.prometheus.prometheusSpec.remoteRead }}
   remoteRead:
 {{ toYaml .Values.prometheus.prometheusSpec.remoteRead | indent 4 }}
